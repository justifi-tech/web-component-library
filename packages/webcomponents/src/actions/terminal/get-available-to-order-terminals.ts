import { ComponentErrorSeverity, Terminal } from '../../api';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetAvailableToOrderTerminals =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchAvailableToOrderTerminals(
        id,
        authToken
      );

      if (!response.error) {
        const terminals =
          response.data?.map((dataItem) => new Terminal(dataItem)) || [];

        onSuccess({ terminals });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        return onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      return onError({
        error: error.message || error,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
