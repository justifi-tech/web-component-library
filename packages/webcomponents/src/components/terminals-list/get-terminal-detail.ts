import { Terminal } from '../../api';
import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetTerminalDetails =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchTerminal(id, authToken);
      if (!response.error) {
        onSuccess({ terminal: new Terminal(response.data) });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      onError({
        error: getErrorMessage(error),
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
  