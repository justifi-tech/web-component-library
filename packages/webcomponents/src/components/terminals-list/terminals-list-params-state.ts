import { createStore } from '@stencil/store';
import { TerminalsQueryParams } from '../../api';

const filterParamsInitialState = {};
const propsParamsInitialState = {};

const filterParamsStore = createStore<TerminalsQueryParams>(() => filterParamsInitialState);
const { state: filterParams, on: onQueryParamsChange } = filterParamsStore;

const propsParamsStore = createStore<TerminalsQueryParams>(() => propsParamsInitialState);
const { state: propsParams } = propsParamsStore;

const getRequestParams = (): TerminalsQueryParams => {
  return {
    ...filterParams,
    ...propsParams
  };
};

const clearParams = () => {
  filterParamsStore.reset();
}

export {
  filterParams,
  propsParams,
  onQueryParamsChange,
  clearParams,
  getRequestParams
};
