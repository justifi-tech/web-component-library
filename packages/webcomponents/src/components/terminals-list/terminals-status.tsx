
export const MapTerminalStatusToBadge = (status: string) => {
  switch (status) {
    case 'connected':
      return "<span class='badge bg-success' title='This terminal is connected.'>Connected</span>";
    case 'disconnected':
      return "<span class='badge bg-danger' title='This terminal is disconnected.'>Disconnected</span>";
    case 'unknown':
      return "<span class='badge bg-secondary' title='The status of this terminal is unknown'>Unknown</span>";
  };
};
