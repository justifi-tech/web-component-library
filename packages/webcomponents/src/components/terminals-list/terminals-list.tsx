import { Component, Event, EventEmitter, h, Prop, State, Watch } from '@stencil/core';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import JustifiAnalytics from '../../api/Analytics';
import { checkPkgVersion } from '../../utils/check-pkg-version';
import { config } from '../../../config';
import { TerminalService } from '../../api/services/terminal.service';
import { makeGetTerminals } from './get-terminals';
import { SubAccountService } from '../../api/services/subaccounts.service';
import { makeGetSubAccounts } from '../../api/get-subaccounts';
import { StyledHost, tableExportedParts } from '../../ui-components';

/**
  * @exportedPart label: Label for inputs
  * @exportedPart input: The input fields
  * @exportedPart input-invalid: Invalid state for inputs  
  * @exportedPart table-head: Table head
  * @exportedPart table-head-row: Head row
  * @exportedPart table-head-cell: Individual head cell
  * @exportedPart table-body: Body of the table
  * @exportedPart table-row: Row of the table
  * @exportedPart table-cell: Individual cell of the table
  * @exportedPart loading-state-cell: Row for loading state
  * @exportedPart loading-state-spinner: Spinner element for loading state
  * @exportedPart error-state: Row for Error state
  * @exportedPart empty-state: Row for Emtpy state
  * @exportedPart pagination: Pagination bar
  * @exportedPart page-item: Pagination button
  * @exportedPart page-link: Pagination link
  * @exportedPart page-link-disabled: Disabled pagination link
*/
@Component({
  tag: 'justifi-terminals-list',
  shadow: true
})

export class TerminalsList {
  @State() getTerminals: Function;
  @State() getSubAccounts: Function;
  @State() errorMessage: string = null;

  @Prop() accountId: string;
  @Prop() authToken: string;
  @Prop() apiOrigin?: string = config.proxyApiOrigin;
  @Prop() columns: string;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  analytics: JustifiAnalytics;

  componentWillLoad() {
    checkPkgVersion();
    this.analytics = new JustifiAnalytics(this);
    this.initializeGetData();
  }

  disconnectedCallback() {
    this.analytics.cleanup();
  }

  @Watch('accountId')
  @Watch('authToken')
  propChanged() {
    this.initializeGetData();
  }

  private initializeGetData() {
    this.initializeGetTerminals();
    this.initializeGetSubAccounts();
  }

  private initializeGetTerminals() {
    if (this.accountId && this.authToken) {
      this.getTerminals = makeGetTerminals({
        id: this.accountId,
        authToken: this.authToken,
        service: new TerminalService(),
        apiOrigin: this.apiOrigin
      });
    } else {
      this.errorMessage = 'Account ID and Auth Token are required';
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: this.errorMessage,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  }

  private initializeGetSubAccounts() {
    if (this.accountId && this.authToken) {
      this.getSubAccounts = makeGetSubAccounts({
        accountId: this.accountId,
        authToken: this.authToken,
        service: new SubAccountService(),
        apiOrigin: this.apiOrigin
      });
    }
  }

  handleErrorEvent = (event) => {
    this.errorMessage = event.detail.message;
    this.errorEvent.emit(event.detail);
  };

  render() {
    return (
      <StyledHost exportparts={tableExportedParts}>
        <terminals-list-core
          getTerminals={this.getTerminals}
          getSubAccounts={this.getSubAccounts}
          onError-event={this.handleErrorEvent}
          columns={this.columns}
        />
      </StyledHost>
    );
  }
};
