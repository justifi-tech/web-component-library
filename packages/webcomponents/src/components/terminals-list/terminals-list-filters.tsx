import { Component, h, Prop } from '@stencil/core';
import { debounce } from 'lodash';
import { filterParams, propsParams, clearParams } from './terminals-list-params-state';
import { StyledHost } from '../../ui-components';
import { ITerminalStatus } from '../../api';

@Component({
  tag: 'justifi-terminals-list-filters',
  shadow: true
})
export class TerminalsListFilters {
  @Prop() terminalId?: string;
  @Prop() terminalStatus?: ITerminalStatus;

  private debouncedSetParamsOnChange: (name: string, value: string) => void;

  componentWillLoad() {
    // debounced input handler for text input
    this.debouncedSetParamsOnChange = debounce(this.setParamsOnChange, 300);

    const propsToSet = {
      terminal_id: this.terminalId,
      status: this.terminalStatus,
    };

    Object.entries(propsToSet).forEach(([key, value]) => {
      if (value) {
        propsParams[key] = value;
      }
    });
  }

  setParamsOnChange = (name: string, value: string) => {
    filterParams[name] = value;
  };

  get terminalStatusOptions(): { label: string, value: ITerminalStatus | '' }[] {
    return [
      { label: 'All', value: '' },
      { label: 'Connected', value: ITerminalStatus.connected },
      { label: 'Disconnected', value: ITerminalStatus.disconnected },
      { label: 'Unknown', value: ITerminalStatus.unknown },
    ]
  }

  render() {
    const filterMenuParams = { ...filterParams }

    return (
      <StyledHost>
        <table-filters-menu params={filterMenuParams} clearParams={clearParams}>
          <div class="grid-cols-2 gap-3 p-1">
            <div class="p-2">
              <form-control-text 
                name="terminal_id"
                label="Terminal ID"
                inputHandler={this.debouncedSetParamsOnChange}
                defaultValue={this.terminalId || filterParams.terminal_id}
                disabled={!!this.terminalId}
              />
            </div>
            <div class="p-2">
              <form-control-select 
                name="status"
                label="Status"
                options={this.terminalStatusOptions}
                inputHandler={this.setParamsOnChange}
                defaultValue={this.terminalStatus || filterParams.status}
                disabled={!!this.terminalStatus}
              />
            </div>
          </div>
        </table-filters-menu>
      </StyledHost>
    );
  }
}
