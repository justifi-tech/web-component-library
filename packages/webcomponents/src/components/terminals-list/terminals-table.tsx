import { h } from '@stencil/core';
import { MapTerminalStatusToBadge } from './terminals-status';

export const defaultColumnsKeys = 'nickname,provider_id,status';

export const terminalTableColumns = {
  nickname: () => (
    <th part="table-head-cell" scope="col" title="The nickname of the terminal">
      Terminal Nickname
    </th>
  ),
  provider_id: () => (
    <th part="table-head-cell" scope="col" title="The provider ID of the terminal">
      Provider ID
    </th>
  ),
  sub_account_name: () => (
    <th part="table-head-cell" scope="col" title="The sub account associated with the terminal">
      Sub Account
    </th>
  ),
  status: () => (
    <th part="table-head-cell" scope="col" title="The current status of each terminal">
      Status
    </th>
  ),
}

export const terminalTableCells = {
  nickname: (value) => (<td>{value}</td>),
  provider_id: (value) => (<td>{value}</td>),
  sub_account_name: (value) => (<td>{value}</td>),
  status: (value) => (<td innerHTML={MapTerminalStatusToBadge(value)}></td>),
}
