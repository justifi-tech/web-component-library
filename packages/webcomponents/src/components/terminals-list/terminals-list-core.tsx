import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { PagingInfo, SubAccount, Terminal, pagingDefaults } from '../../api';
import { TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import { terminalTableColumns, terminalTableCells } from './terminals-table';
import { Table } from '../../utils/table';
import { getRequestParams, onQueryParamsChange } from './terminals-list-params-state';
import { ComponentClickEvent, ComponentErrorEvent } from '../../api/ComponentEvents';
import {
  TableWrapper,
  TableComponent,
  TableHead,
  TableHeadRow,
  TableBody,
  TableRow,
  TableFoot,
  TableFootRow,
  TableFootCell,
  TableClickActions
} from '../../ui-components';

@Component({
  tag: 'terminals-list-core'
})
export class TerminalsListCore {
  @Prop() getTerminals: Function;
  @Prop() getSubAccounts: Function;
  @Prop() columns: string;

  @State() terminals: Terminal[] = [];
  @State() terminalsTable: Table<Terminal>;
  @State() subAccounts: SubAccount[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() pagingParams: any = {};
  
  @Watch('pagingParams')
  @Watch('getTerminals')
  @Watch('getSubAccounts')
  @Watch('columns')
  updateOnPropChange() {
    this.fetchData();
  }

  @Event({ eventName: 'click-event', bubbles: true }) clickEvent: EventEmitter<ComponentClickEvent>;
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentErrorEvent>;

  componentWillLoad() {
    this.terminalsTable = new Table<Terminal>(this.terminals, this.columns, terminalTableColumns, terminalTableCells);
    if (this.getTerminals && this.getSubAccounts) {
      this.fetchData();
    }

    onQueryParamsChange('set', () => {
      this.pagingParams = {};
    });

    onQueryParamsChange('reset', () => {
      this.pagingParams = {};
      this.errorMessage = '';
    });
  }

  fetchData(): void {
    this.loading = true;

    this.getTerminals({
      params: this.terminalParams,
      onSuccess: async ({ terminals, pagingInfo }) => {
        this.terminals = terminals;
        this.paging = pagingInfo;
        this.terminalsTable.collectionData = this.terminals;
        const shouldFetchSubAccounts = this.terminalsTable.columnKeys.includes('sub_account_name');

        if (shouldFetchSubAccounts) {
          await this.fetchSubAccounts();
        } else {
          this.loading = false;
        }

      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  async fetchSubAccounts(): Promise<void> {
    this.getSubAccounts({
      params: this.subAccountParams,
      onSuccess: ({ subAccounts }) => {
        this.subAccounts = subAccounts;
        this.terminals = this.terminals.map((terminal) => {
          terminal.sub_account_name = this.subAccounts.find((subAccount) => subAccount.id === terminal.account_id)?.name;
          return terminal;
        });
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    this.pagingParams = { before_cursor: beforeCursor };
    this.clickEvent.emit({ name: TableClickActions.previous });
  };

  handleClickNext = (afterCursor: string) => {
    this.pagingParams = { after_cursor: afterCursor };
    this.clickEvent.emit({ name: TableClickActions.next });
  };

  rowClickHandler = (e) => {
    const clickedTerminalID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedTerminalID) return;

    const terminalData = this.terminals.find((terminal) => terminal.id === clickedTerminalID);
    this.clickEvent.emit({ name: TableClickActions.row, data: terminalData });
  };

  get terminalParams() {
    const requestParams = getRequestParams();
    const params = { ...requestParams, ...this.pagingParams };
    return params;
  }

  get subAccountParams() {
    let accountIdNumbers = this.terminals.map((terminal) => terminal.account_id);
    let uniqueAccountIds = [...new Set(accountIdNumbers)];
    let accountIdString = uniqueAccountIds.join(',');
    return { sub_account_id: accountIdString };
  }

  get entityId() {
    return this.terminals.map((terminal) => terminal.id);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.terminalsTable.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState && !this.loading;
  }

  render() {
    return (
      <TableWrapper>
        <TableComponent>
          <TableHead>
            <TableHeadRow>
              {this.terminalsTable.columnData.map((column) => column)}
            </TableHeadRow>
          </TableHead>
          <TableBody>
            <TableLoadingState
              columnSpan={this.terminalsTable.columnKeys.length}
              isLoading={this.loading}
            />
            <TableEmptyState
              isEmpty={this.showEmptyState}
              columnSpan={this.terminalsTable.columnKeys.length}
            />
            <TableErrorState
              columnSpan={this.terminalsTable.columnKeys.length}
              errorMessage={this.errorMessage}
            />
            {this.showRowData && this.terminalsTable.rowData.map((data, index) => (
              <TableRow
                data-test-id="table-row"
                data-row-entity-id={this.entityId[index]}
                onClick={this.rowClickHandler}
              >
                {data}
              </TableRow>
            ))}
          </TableBody>
          {this.paging && (
            <TableFoot>
              <TableFootRow>
                <TableFootCell colSpan={this.terminalsTable.columnData.length}>
                  <pagination-menu
                    paging={{
                      ...this.paging,
                      handleClickPrevious: this.handleClickPrevious,
                      handleClickNext: this.handleClickNext,
                    }}
                  />
                </TableFootCell>
              </TableFootRow>
            </TableFoot>
          )}
        </TableComponent>
      </TableWrapper>
    );
  }
}
