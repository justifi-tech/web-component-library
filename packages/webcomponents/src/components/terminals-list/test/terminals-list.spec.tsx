import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { TerminalsList } from "../terminals-list";
import { TerminalsListCore } from "../terminals-list-core";
import { TerminalService } from '../../../api/services/terminal.service';
import JustifiAnalytics from "../../../api/Analytics";
jest.mock('../../../api/services/terminal.service');

beforeEach(() => {
  // Bypass Analytics to avoid errors. Analytics attaches events listeners to HTML elements
  // which are not available in Jest/node environment
  // @ts-ignore
  JustifiAnalytics.prototype.trackCustomEvents = jest.fn();
});

describe('terminals-list', () => {
  it('renders an error message when accountId and authToken are not provided', async () => {
    const page = await newSpecPage({
      components: [TerminalsList, TerminalsListCore],
      template: () => <justifi-terminals-list />,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error message when accountId is not provided', async () => {
    const page = await newSpecPage({
      components: [TerminalsList, TerminalsListCore],
      template: () => <justifi-terminals-list auth-token="abc" />,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error message when authToken is not provided', async () => {
    const page = await newSpecPage({
      components: [TerminalsList, TerminalsListCore],
      template: () => <justifi-terminals-list account-id="abc" />,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('emit an error event when accountId and authToken are not provided', async () => {
    const errorEvent = jest.fn();
    const page = await newSpecPage({
      components: [TerminalsList, TerminalsListCore],
      template: () => <justifi-terminals-list onError-event={errorEvent} />,
    });
    await page.waitForChanges();
    expect(errorEvent).toHaveBeenCalled();
  });

  it('emit an error event when fetch fails', async () => {
    TerminalService.prototype.fetchTerminals = jest.fn().mockRejectedValue(new Error('Fetch error'));

    const errorEvent = jest.fn();
    const page = await newSpecPage({
      components: [TerminalsList, TerminalsListCore],
      template: () => (
        <justifi-terminals-list
          account-id="abc"
          auth-token="abc"
          onError-event={errorEvent}
        />
      ),
    });
    await page.waitForChanges();
    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });
});
