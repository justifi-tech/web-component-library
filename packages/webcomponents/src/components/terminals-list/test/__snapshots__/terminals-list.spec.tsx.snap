// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terminals-list renders an error message when accountId and authToken are not provided 1`] = `
<justifi-terminals-list>
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-terminals-list>
`;

exports[`terminals-list renders an error message when accountId is not provided 1`] = `
<justifi-terminals-list auth-token="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-terminals-list>
`;

exports[`terminals-list renders an error message when authToken is not provided 1`] = `
<justifi-terminals-list account-id="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Account ID and Auth Token are required
      </div>
    </main>
  </template>
</justifi-terminals-list>
`;
